pkg/document/internal_document.go

```diff
@@ -283,34 +295,56 @@ func (d *InternalDocument) ApplyChanges(changes ...*change.Change) ([]DocEvent,
 
 // MyPresence returns the presence of the actor currently editing the document.
 func (d *InternalDocument) MyPresence() innerpresence.Presence {
-	p := d.presences.LoadOrStore(d.changeID.ActorID().String(), innerpresence.NewPresence())
+	if d.status != StatusAttached {
+		return innerpresence.NewPresence()
+	}
+	p := d.presences.Load(d.changeID.ActorID().String())
 	return p.DeepCopy()
 }
 
-// Presences returns the map of presences of the actors currently editing the document.
-func (d *InternalDocument) Presences() *innerpresence.Map {
-	return d.presences
-}
-
-// OnlinePresence returns the presence of the given client. If the client is not
-// online, it returns nil.
-func (d *InternalDocument) OnlinePresence(clientID string) innerpresence.Presence {
+// Presence returns the presence of the given client.
+// If the client is not online, it returns nil.
+func (d *InternalDocument) Presence(clientID string) innerpresence.Presence {
 	if _, ok := d.onlineClients.Load(clientID); !ok {
 		return nil
 	}
 
-	presence, _ := d.presences.Load(clientID)
-	return presence
+	return d.presences.Load(clientID).DeepCopy()
 }
 
-// Presence returns the presence of the given client.
-func (d *InternalDocument) Presence(clientID string) innerpresence.Presence {
-	presence, _ := d.presences.Load(clientID)
-	return presence
+// PresenceForTest returns the presence of the given client
+// regardless of whether the client is online or not.
+func (d *InternalDocument) PresenceForTest(clientID string) innerpresence.Presence {
+	return d.presences.Load(clientID).DeepCopy()
+}
+
+// Presences returns the presence map of online clients.
+func (d *InternalDocument) Presences() map[string]innerpresence.Presence {
+	presences := make(map[string]innerpresence.Presence)
+	d.onlineClients.Range(func(key, value interface{}) bool {
+		p := d.presences.Load(key.(string))
+		if p == nil {
+			return true
+		}
+		presences[key.(string)] = p.DeepCopy()
+		return true
+	})
+	return presences
+}
+
+// AllPresences returns the presence map of all clients
+// regardless of whether the client is online or not.
+func (d *InternalDocument) AllPresences() map[string]innerpresence.Presence {
```

Comment:
Do we need to change `AllPresences` to `PresencesForTest`?

Reply:
I named it `AllPresences` because it is used to retrieve all presences when creating snapshot(for non-testing purposes).

Reply:
It makes sense. 👍